monadic-rx/shared/src/main/scala/mhtml/rx.scala

```diff
@@ -250,6 +262,18 @@ object Rx {
         Cancelable { () => cc.cancel; self.imitating = false }
       } else run(other)(effect)
 
+    case Sharing(self) =>
+      if (!self.isSharing) {
+        self.sharingCancelable = self.impure.foreach{ x =>
+          self.sharingMemo = Option(x)
+          println(s"hello from sharing: $x") // DEBUG
+          effect(x)
+        }
+        self.isSharing = true
+      }
+      else self.sharingMemo.foreach(x => effect(x.asInstanceOf[A]))
+      self.sharingCancelable
```

Comment:
I think you need something more complicated than that. Here if you register twice and cancel the second one it also cancels the first one...

Reply:
Ah, right - some form of "reference counting" (or registration counting) I guess. `Var.foreach` seems to do this, but I should probably look more closely. I was originally trying to use a `Var[A]` instead of memoizing an `A` directly, but didn't think I needed it - maybe I'll try that.